x/concentrated-liquidity/position.go

```diff
@@ -99,20 +99,20 @@ func (k Keeper) HasAnyPositionForPool(ctx sdk.Context, poolId uint64) (bool, err
 	return osmoutils.HasAnyAtPrefix(store, poolPositionKey, parse)
 }
 
-// isPositionOwner returns true if the given positionId is owned by the given sender inside the given pool.
-func (k Keeper) isPositionOwner(ctx sdk.Context, sender sdk.AccAddress, poolId uint64, positionId uint64) (bool, error) {
-	parse := func(bz []byte) (bool, error) {
-		if len(bz) < 1 {
-			return false, fmt.Errorf("insufficient data for parsing boolean")
+func (k Keeper) ensurePositionOwner(ctx sdk.Context, sender sdk.AccAddress, poolId uint64, positionId uint64) error {
+	isOwner := k.isPositionOwner(ctx, sender, poolId, positionId)
+	if !isOwner {
+		return types.NotPositionOwnerError{
+			PositionId: positionId,
+			Address:    sender.String(),
 		}
-		return bz[0] != 0, nil
-	}
-	isOwner, err := osmoutils.HasAnyAtPrefix(ctx.KVStore(k.storeKey), types.KeyAddressPoolIdPositionId(sender, poolId, positionId), parse)
-	if err != nil {
-		return false, err
 	}
+	return nil
+}
 
-	return isOwner, nil
+// isPositionOwner returns true if the given positionId is owned by the given sender inside the given pool.
+func (k Keeper) isPositionOwner(ctx sdk.Context, sender sdk.AccAddress, poolId uint64, positionId uint64) bool {
+	return ctx.KVStore(k.storeKey).Has(types.KeyAddressPoolIdPositionId(sender, poolId, positionId))
```

Comment:
yeah exactly, the problem came from iteration over the prefix. (It would also come from key malleability for anything thats not the final item in the key)

Reply:
Using an iterator here fundamentally didn't make sense, it only makes problems. (You actually would have a key malleability problem if the address wasn't bech32 encoded for instance)

Iterators are also far slower in the SDK, always avoid them if easy to do so.